Embedded ROS 2 Design Page

iluetkeb opened the pull request:
Start of the design page for the ROS 2 Embedded effort.

Includes contributions by all of the OFERA consortium partners.

ralph-lange committed:
Created empty article page on embedded ROS2.

ralph-lange committed:
Added draft for abstract by Ingo Luetkebohle (iluetkeb).

ralph-lange committed:
Very first structure of sections to fill.

ralph-lange committed:
Added first ideas for requirements to list and discuss.

ralph-lange committed:
Link to ofera.eu for detailed requirements document.

ralph-lange committed:
Refined structure and added first descriptions to existing approaches.

BorjaOuterelo committed:
Added proposed architecture diagram.

BorjaOuterelo committed:
Changes svg for png format.

ralph-lange committed:
First draft of the basic requirements.

ralph-lange committed:
Merge branch 'embedded_ROS2' of https://github.com/microROS/design into embedded_ROS2

ralph-lange committed:
Refined list of authors.

ralph-lange committed:
Added link to micro-ROS webpage.

vmayoral committed:
Minor rewording and ROS 2 instead of ROS2

vmayoral committed:
Integrate consistently ROS 2 and ROS 1 approaches

BorjaOuterelo committed:
Updates architecture diagram.
Fixes ARM processor name.

iluetkeb committed:
Merge pull request #1 from microROS/feature/architecture_change

Updates architecture diagram and fixes text typo.

iluetkeb committed:
Fix my name

iluetkeb committed:
Reformat requirements as table.
Call it "wishlist" instead of requirements.

iluetkeb committed:
also mark the items as 'wish' (with a w), instead of 'r' for requirement

iluetkeb committed:
initial questions and actions

iluetkeb committed:
Do not use continuation words in the table.
Split 'MCU-SPEC'.
Fix ANDROID->ARDUINO

iluetkeb committed:
Merge pull request #2 from microROS/embedded_ROS2_tables

Embedded ros2 tables

ralph-lange committed:
Added details on other prior and on-going works.

BorjaOuterelo committed:
Refs #2616. Fixes deliverable link.

ralph-lange committed:
Documented other existing/on-going ROS2-based approaches.

ralph-lange committed:
Changed micro-XRCE-DSS to Micro XRCE-DDS

ralph-lange committed:
Notice on the fact that freeRTPS has obviously been discontinued in 2016.

ralph-lange committed:
Included figure features_with_dependencies.png in embedded ROS 2 article.

ralph-lange committed:
Formulated labels more generic.

BorjaOuterelo committed:
Merge pull request #3 from microROS/feature/deliverable_link

Refs #2616. Fixes deliverable link.

BorjaOuterelo committed:
Refs #3569. Middleware notes:
-Associates seamless integration with middleware questions.
-Adds actions regarding middleware.
-Fixes middleware and protocol names.

BorjaOuterelo committed:
Refs #3569. Extends on some ROS 2 based approaches.

BorjaOuterelo committed:
Refs #3569. Adds Renesas GR-ROSE platform.

iluetkeb committed:
Update articles/embedded.md

Co-Authored-By: BorjaOuterelo <borjaouterelo@gmail.com>

vmayoral committed:
Minor rewording

vmayoral committed:
BLE is beyond our scope, I'd suggest to mention 6LoWPAN instead

BorjaOuterelo committed:
Refs #3569. Resolves pull requests comments:
-Fixes some language misuse.
-Splits sentences to single lines to improve diff readability.

vmayoral committed:
Review wishlist, add boot and infomodel

ralph-lange committed:
Fixed micro-ROS urls.

vmayoral committed:
Minor addition in actions, review ROS 2 prior

ralph-lange committed:
Merge pull request #4 from microROS/embedded_ROS2_on-going_works

Documented prior/on-going works

vmayoral committed:
Address @iluetkeb's request

adamdbrw committed:
Fix typo

Co-Authored-By: vmayoral <victor@erlerobotics.com>

adamdbrw committed:
Rephrasing in INFOMODEL.

Co-Authored-By: vmayoral <victor@erlerobotics.com>

vmayoral committed:
Fix issue with URL

vmayoral committed:
Merge branch 'alr-review1' of https://github.com/microROS/design into alr-review1

Commit:
Merge pull request #6 from microROS/alr-review1

Review from ALR's side, iteration 1

BorjaOuterelo committed:
Refs #3569. Makes the text consistent.

BorjaOuterelo committed:
Merge pull request #5 from microROS/feature/expanddesign

Add notes on ROS2 approaches and add actions

vmayoral committed:
Bold H-ROS prior work to be consistent

iluetkeb committed:
Merge pull request #7 from microROS/alr_review2

Bold H-ROS prior work to be consistent

clalancette commented on articles/embedded.md:
```diff
@@ -0,0 +1,114 @@
+
+Original Author: {{ page.author }}
+
+While standard ROS 2 can run on Linux-class embedded devices, smaller microcontrollers (MCU) are still a challenge. For example, ARM Cortex-M3 and -M4 class MCUs are popular for motion control and sensor integration, but have little RAM and storage. ARM Cortex-M3 is widely used in many sensors due to its optimized power consumption capabilities and ultra low power modes. These microcontrollers ar ...
```
As a matter of style, the design articles generally have one sentence per-line.  This change should be applied throughout the document; I won't mention it elsewhere.

iluetkeb replied:
Will do.

gbiggs replied:
Could you do please that as a priority? I have many comments and if I leave them now, it will be really hard to tell which sentence they apply to... plus they'll be marked as out-dated once you do reformat the document.

iluetkeb replied:
Turns out this is not as trivial as I thought. GFM doesn't handle line breaks in tables well, and we have much of our content in tables. I could do it for the non-table content only quickly, but that wouldn't really help much.

Any suggestions are welcome...

gbiggs replied:
I was going to say you could organise it as definition lists, and I even prepared an example:

```
### W-SEAMLESS

Title
: Seamless integration with ROS&nbsp;2

Description
: The embedded ROS 2 stack shall integrate seamlessly with standard ROS 2 stacks and nodes running on more powerful microprocessors.
  For a standard ROS 2 node, communication (topics, services) from and to software running on the embedded ROS 2 stack should be transparent.
  The same should hold for other core concepts such as parameters, graph introspection, and run-time reconfiguration by the node lifecycle.
```

That's when I discovered that the markdown used in pull request comments is *even less* capable than that used for displaying changed files in pull requests. It would be nice if Github made them all work like Github Pages does. But given the final page will be displayed elsewhere, it probably doesn't matter if it doesn't look nice here.

If that doesn't work for you, then there is also using ever-decreasing sections. 😄 

```
### W-SEAMLESS

#### Title

Seamless integration with ROS&nbsp;2

#### Description

The embedded ROS 2 stack shall integrate seamlessly with standard ROS 2 stacks and nodes running on more powerful microprocessors.
For a standard ROS 2 node, communication (topics, services) from and to software running on the embedded ROS 2 stack should be transparent.
The same should hold for other core concepts such as parameters, graph introspection, and run-time reconfiguration by the node lifecycle.
```

iluetkeb replied:
I went for "ever-decreasing sections" ;-)

clalancette commented on articles/embedded.md:
```diff
@@ -0,0 +1,114 @@
+| Key | Title | Description |
+|-----|-------|-------------|
+| W-SEAMLESS | **Seamless integration with ROS&nbsp;2** | The embedded ROS 2 stack shall integrate seamlessly with standard ROS 2 stacks and nodes running on more powerful microprocessors. For a standard ROS 2 node, communication (topics, services) from and to software running on the embedded ROS 2 stack should be transparent. The same should hold for other core concepts such as parameters, graph ...
+| W-PORTABILITY | **Portability of ROS 2-based software** | The embedded ROS 2 stack shall resemble or directly use the ROS 2 API -- more precisely the rclcpp API -- to facilitate porting standard ROS 2 nodes to MCUs. |
```
When we discussed this at ROSCon, it seemed like we were thinking about targeting more the C-level of API (rcl and friends).  While I'm not that opposed to targeting the C++ API, targeting the C API might make it more palatable to a larger audience.

Actually, thinking about it further, should this be a part of the wishlist/specification at all?  That seems like a detail that might belong in the technical section.

iluetkeb replied:
My intention when calling it "wishlist" has been to distinguish this initial set of ideas/wishes from a fixed set of requirements or a finalized specification. It's just input to be considered, we can always ignore it ;-)

On this specific issue, my view is that the whole C vs C++ issue is a can of worms which I'd like to avoid by being open to both, and then looking who's actually willing to contribute code, and in what language. Right now, we see a mix.

Maybe we should mention both rclc and rclcpp here as well and be done with it?

clalancette replied:
> Maybe we should mention both rclc and rclcpp here as well and be done with it?

Either clarifying like you mention (or removing a specific callout) would both be fine by me.

clalancette commented on articles/embedded.md:
```diff
@@ -0,0 +1,114 @@
+| W-PORTABILITY | **Portability of ROS 2-based software** | The embedded ROS 2 stack shall resemble or directly use the ROS 2 API -- more precisely the rclcpp API -- to facilitate porting standard ROS 2 nodes to MCUs. |
+| W-DEVICES | **Support of a broad spectrum of device classes** | The embedded ROS 2 stack shall support a broad range of MCU device classes, starting from a few tens kilobyte of RAM. |
+| W-RTOS | **Support of a broad spectrum of RTOS(s)** | The stack shall be adaptable to different RTOS (e.g., NuttX, FreeRTOS, Zephyr) and provide abstractions with regard to the application software to facilitate porting application components to between RTOSs.|
+| W-COMM | **Support of prevalent communication technologies** | To cover the broad range of use-cases for MCUs in robotics, the embedded ROS 2 stack shall be usable with the default ROS 2 middleware standard DDS, simple (custom) serial communication protocols just as common wireless technologies like 6LoWPAN |
```
The previous line is a bit unclear to me.  Does this mean the embedded ROS2 stack will be usable with DDS *and* simple serial communication protocols *and* 6LoWPAN?  A bit of clarification on that line would help.

iluetkeb replied:
Again, it's a wish only, but I would say that when targeting battery-powered devices, plain old IPv4 isn't gonna cut it, and when targeting MCU to MCU communication on a single board, XRCE-DDS is overkill.

I think one of the ideas would be to hide this in the XRCE-DDS implementation, by specifying profiles. micro-XRCE by eprosima already runs over serial lines as well as UDP.

clalancette commented on articles/embedded.md:
```diff
@@ -0,0 +1,114 @@
+| W-STATIC | **Use static initialization** | Static initialization is less error-prone, easier to analyze, and shifts memory usage from RAM to flash. It is a requirement for going to very small resource use. |
+| W-BOOT | **Quick boot times compared to native ROS 2 machines** | Microcontrollers provide with the capability of booting very quickly (typically, in the order of tenths of milliseconds or less). This  enhances the existing landscape of ROS 2-native devices which in the best cases, require a few seconds to boot an embedded Linux.|
+| W-INFOMODEL | **A common interface that facilitates interoperability across different vendors** | An information model for robot devices, adapted by vendors of robot modules, would lower costs of integration of robotic systems. Within the OFERA project we propose and use the [Hardware Robot Information Model (HRIM)](https://acutronicrobotics.com/modularity/hrim/).|
+
```
In the entire preceding list, I see some of the terms that I would expect from a specification, such as "shall", "should", etc.  However, they aren't consistently applied and they aren't defined up-front.  Should we formalize this along the lines of:

* "shall" - A must-have requirement, the software would not be considered suitable unless it met this.
* "should" - A nice-to-have requirement, the software could operate without it but would be better with it

iluetkeb replied:
This particular three items don't have any of that -- why did you highlight it here?

Anyway, this is an oversight of editing. The earlier parts of the list came from a requirements document and I didn't take the shalls out. These three items were added later and don't use that kind of language.

clalancette replied:
> This particular three items don't have any of that -- why did you highlight it here?

I was just reading the list and put it at the end.

> 
> Anyway, this is an oversight of editing. The earlier parts of the list came from a requirements document and I didn't take the shalls out. These three items were added later and don't use that kind of language.

OK, in that case, it's fine for now.

clalancette commented on articles/embedded.md:
```diff
@@ -0,0 +1,114 @@
+
+## Questions
+
+To determine how feasible that is, and to come up with the actual requirements and/or design, a few questions come to mind. I've cross-referenced them to the relevant wishlist items (most of the there's more than one -- this suggests areas where trade-offs may be required, or alternatives pursued).
```
`most of the there's more than one` -> `in most cases there are more than one`, maybe?

Also, as a small matter of style, we should always write in the third-person (avoid `I've` and similar).  Maybe `I've cross-referenced them` -> `The questions have been cross-referenced`

clalancette reviewed (commented):
In the diagrams, it might be a good idea to define the acronyms (I'm pretty sure AL stands for "Abstraction Layer", for instance, but it would be better to be sure).  Also, at the very top of the micro-ROS_proposed_architecture.png file, it is unclear to me what "micro" means alongside "tf" and "diagnostics".

I haven't yet had time to think more deeply about the Design section; I'll try to do that, though it may have to wait until next week.  I'll leave these early comments for now.

iluetkeb commented:
@clalancette I see that the action's PR is on a branch in this repo. Maybe more people can add commits that way? If yes, maybe we could re-target this to a branch, merge it there without discussion, and then do a PR from that branch to gh2-pages and re-commence the discussion on that? Just let me know.

iluetkeb committed:
Update embedded.md

Reformatted wishlist table into sections, to comply with OSRF's "short lines" requirement.

iluetkeb committed:
Reformatted non-table parts to comply with short-line requirement.

iluetkeb committed:
Final set of reformattings to achieve short lines.

iluetkeb commented:
@clalancette I saw that it is easily possible to re-target the PR to a different branch. If you guys would be willing to give me access to the design repo, I'd very much prefer creating an "embedded_ROS2" branch there and continue the discussion in that.

BorjaOuterelo committed:
Refs #3512. Adds references to new ROS 2 integration initiatives.

BorjaOuterelo committed:
Refs #3512. Joins ros2arduino to related Robotis initiative.

clalancette commented:
> @clalancette I saw that it is easily possible to re-target the PR to a different branch. If you guys would be willing to give me access to the design repo, I'd very much prefer creating an "embedded_ROS2" branch there and continue the discussion in that.

I'll check what our policy with that is.  For now, I'll just pull this PR and push directly to a branch on this repo.

clalancette commented:
All right, closing this in favor of https://github.com/ros2/design/pull/198

gbiggs commented on articles/embedded.md:
```diff
@@ -0,0 +1,309 @@
+power modes. These microcontrollers are often running with small batteries, 
+e.g. for environmental sensors or small consumer robots, which requires
+efficient power management. We aim to address these devices through a 
+combination of specialized, ROS 2-interoperable stack as well as the use
```
Show your reasoning for *not* addressing these devices through improvements in rmw/rcl.

iluetkeb replied:
At this stage "specialized" doesn't mean it couldn't be rmw/rcl, but also it doesn't *have* to be rmw/rcl at all costs. What is clear, is that it should be interoperable -- that, after all, is the main goal.

gbiggs commented on articles/embedded.md:
```diff
@@ -0,0 +1,309 @@
+### W-PORTABILITY: *Portability of ROS 2-based software*
+
+The embedded ROS 2 stack shall resemble or directly use the ROS 2 API 
+-- more precisely the rclcpp API -- to facilitate porting standard 
```
This implies you want to have a C++ API but find the current rclcpp implementation too heavy for tiny scale devices. Is that why you are creating a separate implementation? If so, why not improve rmw/rcl and then create a separate C++ client library built on top of them, e.g. rclcpp_tiny? This would preserve the benefits of everyone wrapping rcl, bring the improved performance benefits of this project to all client libraries, and still satisfy your desire for a C++ API implementation oriented to low-resource systems.

iluetkeb replied:
This implies merely that we haven't made up our mind, yet, on what it's going to be.

gbiggs commented on articles/embedded.md:
```diff
@@ -0,0 +1,309 @@
+### W-DEVICES *Support of a broad spectrum of device classes*
+
+The embedded ROS 2 stack shall support a broad range of MCU device 
+classes, starting from a few tens kilobyte of RAM.
```
And going to where? If rcl can be made to support something as low as a few tens of kilobytes of RAM then it probably already supports the other end of your range.

I also note that the RFP for DDS-XRCE, which you plan to use, states that it targets a minimum RAM of 100KB. I'm aware that the eProsima implementation can go smaller than this, but other implementations might not.

iluetkeb replied:
This is just a wish. I'm very much interested in addressing very tiny devices, such as small Arduinos or the like. Whether rcl can support these remains to be seen.

btw, DDS-XRCE is the default option, but, like in "regular" ROS 2, it should be possible to replace it with other, possibly smaller, middleware options.

gbiggs commented on articles/embedded.md:
```diff
@@ -0,0 +1,309 @@
+
+The stack shall be adaptable to different RTOS (e.g., NuttX, FreeRTOS, 
+Zephyr) and provide abstractions with regard to the application 
+software to facilitate porting application components to between RTOSs.
```
How do you plan to deal with non-POSIX RTOSs?

iluetkeb replied:
In the first instance, through an abstraction layer, as shown in the architecture diagramm.

In the future, we might also be able to have some of this through specialized executors, for example (given that scheduling is one of the areas where there are major differences).

gbiggs replied:
Big :+1: to providing specialised executors for specific RTOS's. OpenRTM used that approach and it worked *really* well.

Custom executors should also be able to provide support for advanced scheduling concepts, especially on bare-bones systems or RTOSs without powerful scheduling of their own, and also for systems where we want absolute guarantees of scheduling.

gbiggs commented on articles/embedded.md:
```diff
@@ -0,0 +1,309 @@
+To cover the broad range of use-cases for MCUs in robotics, the embedded 
+ROS 2 stack shall be usable with the default ROS 2 middleware standard 
+DDS, simple (custom) serial communication protocols just as common 
+wireless technologies like 6LoWPAN
```
These all sound wonderful to have in rmw/rcl. I don't think this creates a need for a separate implementation. Why can't you create rmw implementations for these different communications methods, or use a DDS or DDS-XRCE implementation that can use them as its wire layer? The latter would mean that the ROS2 stack never even has to know that 6LoWPAN is being used. I know that RTI's DDS can support things like serial communications.

iluetkeb replied:
Micro DDS-XRCE already supports serial as well, and its fairly transparent, except at initialization time.

Open issues in this regard include things such as re-use of buffers (and zero-copy in general).

BorjaOuterelo replied:
Currently here at eProsima have on our roadmap some changes regarding buffering. A design phase should start shortly.

gbiggs commented on articles/embedded.md:
```diff
@@ -0,0 +1,309 @@
+adaptation to specific hardware features and different RTOS, but also to
+allow the integration with existing frameworks and to be able to create 
+customized sub-stacks. For example, it shall be usable in the style of 
+rosserial, thus bringing basically the message serialization to the MCU 
```
If there are use cases that are only satisfiable using a rosserial-style approach, then I think a better approach is to create a rosserial2 for those and improve the existing ROS2 stack for the rest.

gbiggs commented on articles/embedded.md:
```diff
@@ -0,0 +1,309 @@
+customized sub-stacks. For example, it shall be usable in the style of 
+rosserial, thus bringing basically the message serialization to the MCU 
+only. Similarly, it should be possible to derive a sub-stack that provides 
+node graph features and publish/subscribe but no parameters and services, 
```
This seems doable with rcl.

Enabling developers to save resources "by picking only those features they really need" is something that would be useful at any scale. It would need some modifications to rmw/rcl to support there, for example, changes in library structure, or using compile-time flags to enable/disable features. The potential impact of these changes needs to be analysed. rcl already has separate libraries for lifecyle nodes and actions and while I don't like the current layout of the APIs in regards to this separation, structurally this is clearly possible.

iluetkeb replied:
On embedded, we don't have to have compile-time flags for that, because linking can often be static, and then the linker will only pull in the functions which are actually referenced.

I'm not sure we want this for the shared library case, but if there are use cases, that would be a separate discussion, IMHO.

gbiggs commented on articles/embedded.md:
```diff
@@ -0,0 +1,309 @@
+### W-CONTROL: *Support control-oriented applications*  
+
+MCUs are great for control, and micro-ROS should be as well. This usually 
+means hard real-time performance with low jitter and small response times.
```
Isn't this already a goal for rcl, and at least partly achieved? And it's a desirable feature at any scale.

iluetkeb replied:
This goes across the stack. Current issues are:
 1. The use of locking (usually with mutexes) in some middleware implementations
 1. The allocation of memory at runtime

The first is not an rcl issue as such, but important nonetheless. The second *could* be addressed using allocation from a static block (which is what rcl currently suggests, AFAICT), but it is not clear this is an ideal approach. It definitely leads to more error handling than doing this at compile-time.

gbiggs commented on articles/embedded.md:
```diff
@@ -0,0 +1,309 @@
+
+MCUs are often used in battery-powered applications, and/or in applications
+with a large amount of standby time. Therefore, the stack should make it 
+easily possible to save power.
```
Even if I have a reasonably large processor in my robot, I may still want it to be capable of using low power modes and generally be power-efficient. A scifi humanoid robot butler would spend a lot of time standing around and should do so in low-power mode.

iluetkeb replied:
Exactly. One common use case would be robotic vacuums -- they typically move only one hour per day, and spend the rest idling.

The thing is, on the high level we can often turn the whole processor off indefinitely. This makes it fairly easy to address. On the MCU side, we need to keep at least some functions running, e.g. those monitoring the battery.

gbiggs commented on articles/embedded.md:
```diff
@@ -0,0 +1,309 @@
+
+Static initialization is less error-prone, easier to analyze, and shifts 
+memory usage from RAM to flash. It is a requirement for going to very small 
+resource use. 
```
It's also a requirement for safety-critical systems that are large. This item is good to be able to do in rmw/rcl as well. It would be better to enable it there than create a separate software stack.

It's worth adding a note here that static initialisation extends to initialising the entire distributed system with a known layout of nodes, topics, etc. that does not change at run time.

iluetkeb replied:
> It's worth adding a note here that static initialisation extends to initialising the entire distributed system with a known layout of nodes, topics, etc. that does not change at run time.

Yes, that's exactly what we mean here, and that's a big change for rcl, because currently it expects to allocate these at runtime.

In our embedded prototypes, we do this using preprocessor macros. Other systems also code generation for this part.

Maybe I should add some more text in the document, explaining why that is difficult to do with rcl/rmw currently, and what other options we have.

gbiggs replied:
That's exactly the sort of information we need.

gbiggs commented on articles/embedded.md:
```diff
@@ -0,0 +1,309 @@
+Microcontrollers provide with the capability of booting very quickly
+(typically, in the order of tenths of milliseconds or less). This enhances
+the existing landscape of ROS 2-native devices which in the best cases, 
+require a few seconds to boot an embedded Linux.
```
Any improvements to initialisation times for the existing ROS2 stack are going to be welcomed by anyone working at any scale.

iluetkeb replied:
It is important to know where this comes from. Some of the main ingredients are:
 1. Know your hardware at compile time (so, no probing for addresses, no trying unused drivers)
 1. Compile everything statically (so, no symbol remapping at load time)
 1. No use of hardware which takes long to initialize

And then there is making sure that each node's initialization method runs fast, which is not a framework issue as such.

This is just to say: Don't expect miracle from this work on your Linux box ;-)

gbiggs commented on articles/embedded.md:
```diff
@@ -0,0 +1,309 @@
+An information model for robot devices, adapted by vendors of robot modules, 
+would lower costs of integration of robotic systems. Within the OFERA project 
+we propose and use the 
+[Hardware Robot Information Model (HRIM)](https://acutronicrobotics.com/modularity/hrim/).|
```
Something that all of ROS2 should be using.

gbiggs commented on articles/embedded.md:
```diff
@@ -0,0 +1,309 @@
+
+#### Related: W-RTOS, W-PORTABILITY
+
+### Q-BUILD: *How do we handle the RTOS(s) respective build-systems?*
```
Can plugins for colcon support them? Although I know that there are many embedded devices with annoying Windows-only tool chains.

iluetkeb replied:
Not sure. The general issue is that you might need headers from the RTOS, and those can come with a lot of preprocessor magic set by their build systems. It *may* be possible to create a config.h or similar, but this remains to be investigated.

Here's what I know:
 * For FreeRTOS, you usually include it in the app, so no issue there
 * For Arduino, you could invoke Arduino's build_platfrom from colcon, but it expects some stuff to be *copied* to the Arduino workspace, so you at least need to know needs to be copied there
 * NuttX has its own build-system similar (or even the same, not sure) as the Linux one. It creates a config.h and then uses a regular Makefile infrastructure to compile. It also expects applications to reside in an "apps" directory. There's also some dynamic loading capability, but not sure how that works.
 * Zephyr uses CMake, so should be easy to integrated
 * Not sure about RIOT, @vmayoral might know more.

gbiggs commented on articles/embedded.md:
```diff
@@ -0,0 +1,309 @@
+
+#### Related: W-RTOS, W-PORTABILITY 
+
+### Q-LANG: *Which language(s) should be used, and at what spec level?*
```
At least in Japan, C is still the dominant language for embedded, even though C++ is perfectly usable. More of a problem is that embedded device compilers tend to be really far behind the standards.

iluetkeb replied:
For most of the devices we target, there's a gcc version, so we can compile C++. However, we do not always have a libstdc++, and if there is one, it can be old.

Currently, @BorjaOuterelo is trying out libcxx from the LLVM ecosystem.

gbiggs commented on articles/embedded.md:
```diff
@@ -0,0 +1,309 @@
+
+#### Related: W-PORTABILITY, W-MODULARITY, W-STATIC, W-POWER
+
+### Q-PERF: * What are the performance implications of the API?*
```
If there are performance problems in the rmw or rcl APIs, then I'd like to see those fixed there rather than making a new implementation with a different API.

iluetkeb replied:
There are trade-offs here. For example, if you do everything at runtime, you're flexible, but often slower and more error-prone. Just saying...

gbiggs commented on articles/embedded.md:
```diff
@@ -0,0 +1,309 @@
+others have already undertaken or are planning exploratory work.
+
+*Meta-Note*: Please only add a short description here, linking to 
+more detailed pages if necessary.
```
You also need at least a short description of the results/conclusions of these experimental actions.

gbiggs commented on articles/embedded.md:
A CMake generator for each RTOS platform?

iluetkeb replied:
CMake generators are written in C++, and not so trivial...

At ROSCon, @vmayoral and team have presented an approach for this, which is based on transforming CMake using CMake ;-)

gbiggs commented on articles/embedded.md:
Can we fix the need for manual changes?

iluetkeb replied:
This is related to the build-system issue. Once we got that covered, this goes away.

gbiggs commented on articles/embedded.md:
This is the question that should have been asked and answered before the decision was made to create a separate implementation of the ROS2 stack (and not just for rclcpp but for all parts of the stack). I'm not sure why it's only just being asked now.

iluetkeb replied:
Which decision are you talking about? This is a PR for a design document, and at this stage it very purposefully only lists *wishes* -- neither fixed *requirements* nor decisions.

gbiggs commented on articles/embedded.md:
So create an rmw implementation for it so the entire ROS2 stack benefits.

iluetkeb replied:
This has been done eProsima.

gbiggs commented on articles/embedded.md:
I would love to have an rmw implementation for this, rather than seeing it only used in a separate implementation of the ROS2 libraries.

iluetkeb replied:
See above.

gbiggs commented on articles/embedded.md:
Strange that the project documents and this document don't mention this investigation up front and instead start from the assumption of needing to create a separate, "compatible" implementation of the core ROS2 libraries. I'd like to see the outcome of this work and in particular numbers showing that it's just not possible to get rmw and rcl down to the tiny scale embedded size before I agree that *the* approach for ROS2 for these devices should be a separate implementation of the core libraries.

iluetkeb replied:
Can we resolve this?

gbiggs commented on articles/embedded.md:
This looks like a great starting point for talking about how to improve the structure of rcl to allow major features to be turned on and off at compile time.

gbiggs commented:
It looks like the discussion is going to move back here, so I'll leave my comments on this pull request.

First, some general comments. I read both this document and the linked OFERA report.

The biggest problem I have is that there is an inherent assumption right from the start that something else needs to be created to support small-scale embedded devices, i.e. rmw/rcl cannot be used directly even with modifications. I haven't seen any evidence to support this assumption and without seeing some I can't agree with the direction this document proposes for how ROS2 will support small-scale devices. I would prefer to see rmw/rcl used directly and modified where necessary, unless it can be shown that making them work at a small scale will compromise them for medium and large scales.

Not using rcl directly does two things that I don't like:
* It creates a separate implementation of core ROS2 functionality, the urcl library, which undermines the concept that rcl provides the core functionality and all other client libraries just wrap rcl, meaning every client library gets exactly the same behaviour and receive any changes and fixes together.
* It requires a bridge to talk between "normal" ROS2 and embedded ROS2 nodes.

For the linked report's statements and requirements, I have some additional comments relevant to the above:
* The statement that micro-ROS will split functionality into separate libraries to enable developers saving resources "by picking only those features they really need" is something that would be useful at any scale, and I see no reason rmw/rcl cannot enable this in some way (changes in library structure, or using compile-time flags to enable/disable features, for example). rcl already has separate libraries for lifecyle nodes and actions and while I don't like the current layout of the APIs in regards to this separation, structurally this is clearly possible.
* The micro-ROS APIs for things like life cycles, predictable scheduling and system modes are apparently going to be much richer than those in rcl due to "advanced concepts specific to micro-ROS". What are these advanced concepts and why are they specific to tiny embedded devices? For example, I think that rich control over predictable scheduling is something we will be very interested in for a real time system at any scale. I'm sure Dejan and his people at Apex.AI would agree with me and they are working at the level of autonomous cars.
* There are many statements saying things like "create a generic framework in the spirit of ROS" and "be ROS2 compatible". These imply that the OFERA goal is to create something compatible with ROS2, which makes me think "then why is this design document going on ROS2's design site?" It's probably just wording for a project report, but I still think that it sounds weird.
* All the things listed in the table in 4.2, such as allowing static node and topic layouts and support for sleep states are things that are desirable in rcl as well.
* All of the performance requirements listed in section 5.2 are desirable at any scale. Even in a large system, I still want things like rapid start-up times, no-copy communication within the same MCU, and minimal power usage. It therefore makes more sense to me that rmw/rcl be improved to satisfy these requirements rather than putting that effort into a separate implementation. Some of the requirements are already satisfied by rcl.

So to sum up, if someone can show a compelling argument why rmw/rcl cannot satisfy the very small scale devices use case that this document targets, then I can agree with the need for a separate implementation. But so far I have only seen this as an unsupported assumption, and the listed requirements seem reasonable for rmw/rcl. Similarly, the wishlist are all things that I would want at larger scales as well.

iluetkeb commented:
@gbiggs I think we got off a bit on the wrong foot here, most likely because you recall an earlier discussion on discourse whether rcl could be re-used or whether there's going to be a  ucrlc. Because, really, it has not been my intention to convey that there is a decision to do a ucrlc. If there's text in this document that says otherwise please point it out (and I've taken note on the sentence with "specialized" above).

That said, there is observable evidence from existing embedded implementations for small and tiny devices that should give us pause on whether barging ahead and porting rmw/rcl at all costs is the right approach. That's why this document is using fuzzy language that postpones such a decision and instead commits to measuring it.

FWIW, right now, as part of the OFERA project, eprosima has provided an rmw implementation for Micro XRCE-DDS. They also ported rcl, and there is ongoing work to port rclcpp. We have identified several issues with that (e.g., related to 64bit atomics) and are talking to the OSRF about it. @BorjaOuterelo can probably provide more info.

In addition to that, several people, including Robotis, Amazon, and myself, have also pursued alternatives to that, for various reasons. You may not like it, but I would argue that you actually should, because these alternatives provide concrete examples of how to do things differently, which IMHO is a much better foundation for discussion than "what if's".
**Update:** and this also allows us to show something desirable without having to figure out to reconcile it with sometimes conflicting current design decisions in rmw/rcl. Because, for example, things like static allocation are not only about allocating all the memory at the beginning, but also about being able to make the assumption that structures are going to be valid without having to check it again over and over. And that's not going to fit in very well with the way rcl does things currently. 

This is ongoing work, and benchmarking results will trickle in, so that we get more evidence for decision-making.

btw, let me also note that this is not only about resource use, but also about how to get rid of the plethora of checks in rcl and rmw on whether a pointer is still valid ,-) I very much hope that at least some of that will get into rcl proper :-)

gbiggs commented:
> it has not been my intention to convey that there is a decision to do a ucrlc. If there's text in this document that says otherwise please point it out (and I've taken note on the sentence with "specialized" above).

This document is more ambiguous about the intended approach. The linked OFERA report is not. It seems quire clear in that report that the intention is to create a separate software stack that is ROS2-compatible to meet the requirements. There are statements like:
* "The next layer is the micro-ROS client library (urcl) analgously to the ROS client library (rcl) in ROS 2."
* "Since the micro-ROS aims to create a generic framework for robotics in the spirit of ROS"

and even a *requirement*:
* "Micro-ROS bridge: functionalities that are specific to the bridge between ROS2 and micro-ROS."

I don't know how I can read these any other way than that OFERA has decided from the outset that a separate software stack will be created. It's making this decision with any evidence for why it is necessary that bugs me. Of course I have no input into OFERA so if that project wants to do that, then it can go right ahead. However, when it comes for how *ROS2* is going to define its tiny-scale embedded support, then I think more evidence needs to be provided to support such a decision. The problem comes from using that report as support for this design document, and this document also being written in a way that implies creating a separate software stack is the goal due to statements like the one near the very start that talks about creating a ROS2-interoperable stack. Perhaps such statements could be rephrased to include reusing the existing client libraries as much as possible, and making modifications where possible, and using as little custom code as possible?

> several people, including Robotis, Amazon, and myself, have also pursued alternatives to that, for various reasons. You may not like it, but I would argue that you actually should, because these alternatives provide concrete examples of how to do things differently, which IMHO is a much better foundation for discussion than "what if's".

I don't think I said I don't like that work being done. If it seems like I did, then that was not my intention and I apologise for the confusion. I want to see these approaches tried and I want to see using rmw/rcl as much as possible tried, so we have numbers that can be used to make a decision. I even said as much in one of my previous comments.

iluetkeb commented:
Update: I had some more text here earlier, but it's really besides the point.

In the OFERA project, we're hedging our bets with regard to how much we can re-use, because there's some technical and organizational uncertainty. Please do not overinterpret this, and instead lets move on to the technical discussion. If we didn't want to play by the communities terms, we wouldn't be here.

> "Micro-ROS bridge: functionalities that are specific to the bridge between ROS2 and micro-ROS."

That's mainly the XRCE-DDS agent. Because we're not using the same middleware, we have to bridge. People have tried to bring DDS to small devices, and it didn't work. That's not just a matter of memory, it's also about enabling power-saving etc. (which doesn't go well with a protocol that assumes you're always listening).

Things which might distinguish this bridge from the "plain" XRCE DDS agent include ROS-specific things, such as TF filtering, etc.

iluetkeb committed:
Remove wording which might imply that re-implementing rmw/rcl is a goal

gbiggs commented:
I think that the changes made in https://github.com/ros2/design/pull/197/commits/2b7cb6c08afad7a024e2710f97c083faf89d2e0d address my concerns about the planned direction for the work.

smorita-esol commented:
Why don't you discuss the rmw (or rcl) plugging multiple middlewares?

I think the embedded ROS aims to support non-DDS or non-XRCE DDS middlewares.
And there are some cases that multiple middlewares exist in one system (e.g. standard DDS and MQTT).
In those cases, at least one node straddling over plural middlewares is needed to subscribe the one side's topic (standard DDS) and process it, publish the processed topic to the other side (MQTT), or vice versa.

This function is apparently needed in the embedded system where suitable middlewares and/or wire protocols tend to vary for each purpose.
But, it is also beneficial for non-embedded systems. For example, SONY stated that they are considering if their middleware is suitable for ros2 or not (https://roscon.ros.org/2018/presentations/ROSCon2018_Aibo.pdf, p29). ROS2 developer like SONY will probably implement the function above to harmonize their middleware with the standard DDS.

iluetkeb commented:
> Why don't you discuss the rmw (or rcl) plugging multiple middlewares?

That's not something that came up during our discussions so far. 

> This function is apparently needed in the embedded system where suitable middlewares and/or wire protocols tend to vary for each purpose.

In the embedded systems that I work with, such decisions are made at compile time.

iluetkeb commented:
> Why don't you discuss the rmw (or rcl) plugging multiple middlewares?

This would be an RCL topic in general, wouldn't it?

In my experience, for MCUs, this is decided at deployment time. Do you have use cases where it's decided at runtime?

> In those cases, at least one node straddling over plural middlewares is needed to subscribe the one side's topic (standard DDS) and process it, publish the processed topic to the other side (MQTT), or vice versa.

In our systems, we have an agent that does such things. It is running on the Linux side. This is not something we burden then MCU with.

Maybe I should add a system sketch early on in the document, to show the overall system architecture and eco-system.

iluetkeb committed:
Merge pull request #8 from microROS/feature/addnewros2

Adds new projects.

smorita-esol commented:
Thanks for your comments.

>This would be an RCL topic in general, wouldn't it?

Currently, I don't have any idea about which layer (rmw or rcl) is better to add the function.

>In my experience, for MCUs, this is decided at deployment time. Do you have use cases where it's decided at runtime?

To avoid misunderstanding, I'd like to mention that those plural middlewares have to be used concurrently (not exclusively) in runtime. Of course, the set of middlewares to be used in runtime has to be decided until deployment time.

> In our systems, we have an agent that does such things. It is running on the Linux side. This is not something we burden then MCU with.

It may be true if we choose the set of standard DDS and XRCE DDS. But, there are some middlewares which have no functions of the agent publishing/subscrinbing standard DDS topic. 
If we accept those middlewares, we should provide the function to bridge over different middlewares without agent.
Once we provide the function, we can also choose the system configuration where the system has no DDS middlewares(e.g. combination of ZeroMQ and EtherCAT).

iluetkeb commented:
@clalancette What's your view on this? Can we merge?

AFAICT, we resolved all major issues with this version of the document. We also had a SIG telco a while ago where we agreed on the further direction (i.e., sticking with stock rcl/rmw).

Of course, as design documents go, it definitely needs further work and we're going to further develop it as we go along. However, since there have been no more issues identified, I think this version could go in as a baseline for now.

iluetkeb commented:
@gbiggs would you agree that we have addressed your points? You made one comment in this direction, but I'm not sure whether the PR as a whole has been marked valid.

dirk-thomas commented on articles/embedded.md:
```diff
@@ -0,0 +1,318 @@
+## Important Differences / Assumptions
+
+One of the major difference for the embedded stack is to be able to
+run not (only) on Linux, but also on Real-Time Operating Systems, or
```
I don't see a reason why the existing ROS 2 code couldn't run on a Real-Time Operating Systems? I guess this is aiming for a very specific sub-category of OSes. Maybe it makes sense to state them here explicitly.

iluetkeb replied:
The answer to this depends on what you mean by "existing ROS 2".  See our design diagram for the layers we expect to be able to re-use.

In more detail:
 * We expect that RCL and RMW can run on RTOS's with POSIX APIs when targeting devices with at least 32kB of RAM. Some modifications might be useful to make them run better, but that's a different topic.
 * DDS *cannot* run on most target devices, as is well known, hence the use of XRCE-DDS. This then also excludes the corresponding rmw_ and type_support_ layers implementations. We did rmw_ and type_support_ implementations based on Micro-XRCE-DDS, however, so in principle this runs.
 * rclcpp is tricky, because of the C++ standard library, which is large and not trivial to port. Borja from eProsima did some experiments with libcxx and it doesn't look too bad, but there's also a few issues. Moreover, it might take so much FLASH and RAM that most people want to avoid it. This is something where we still need more benchmarks. An alternative might be  rclc.

After that, we get into the realm of libraries above rclcpp. 
 * TF is important, and I've already done a bit of work on that, with more to come in early 2019. Most likely I will produce a stripped down 'C' version, so that we don't require rclcpp just for TF.
 * Executors could also benefit from specific support, as mentioned above

traversaro replied:
Regarding TF on embedded systems, have you checked the https://github.com/ESROCOS/tools-transformer library from the ESROCOS (https://www.h2020-esrocos.eu/) European project?

gavanderhoorn replied:
@traversaro: that seems to be C++ based?

iluetkeb replied:
@traversaro thanks for the hint, I wasn't aware of that but will check it out. are you involved with the project, so that I could ask you about it?

@gavanderhoorn C++ is not strictly impossible, most embedded compilers support it. what's more problematic is the standard library. but yeah, it's always something to look critically at, unfortunately (I usually prefer C++ over 'C' due to its better consistency features, but have had to accept that it can be an issue).

traversaro replied:
> @traversaro thanks for the hint, I wasn't aware of that but will check it out. are you involved with the project, so that I could ask you about it?

I  am not involved with the project at all, but I learned a lot about it by reading the deliverables, that are detailed but readable documents: https://www.h2020-esrocos.eu/publications/deliverable-documents/ .

> @traversaro: that seems to be C++ based?

It is indeed C++, but if I remember correctly from the deliverable is meant to run on the top of RTEMS or similar OS, so I think it is still worth checking it out.

dirk-thomas commented on articles/embedded.md:
```diff
@@ -0,0 +1,318 @@
+
+To cover the broad range of use-cases for MCUs in robotics, the embedded 
+ROS 2 stack shall be usable with the default ROS 2 middleware standard 
+DDS, simple (custom) serial communication protocols just as common 
```
Imo the statement "usable with" is not specific enough in terms of goals. In the meaning of "compatible with on the wire" there is simple no other choice than DDS. In the meaning of "compatible using e.g. a bridge" I don't see much value in mentioning it - that part is explicitly covered by `W-SEAMLESS`. Technically even ROS 1 is compatible in that sense using the `ros1_bridge`.

iluetkeb replied:
What's meant here is "usable" in the same way that you can exchange rmw_-implementations now.

Not sure how to rephrase that so it becomes clearer. Maybe just mentioning rmw would do the trick? Other suggestions welcome.

gbiggs commented:
> @gbiggs would you agree that we have addressed your points? You made one comment in this direction, but I'm not sure whether the PR as a whole has been marked valid.

Yeah, I guess so. From your recent updates it sounds like you are going in the direction I prefer.
